pkg/collections.go

```diff
@@ -0,0 +1,308 @@
+				}); err != nil {
+					return err
+				}
+			case driver.PersistentIndex:
+				if _, _, err := destinationColl.EnsurePersistentIndex(ctx, index.Fields(), &driver.EnsurePersistentIndexOptions{
+					Name:   index.UserName(),
+					Unique: index.Unique(),
+					Sparse: index.Sparse(),
+				}); err != nil {
+					return err
+				}
+			case driver.SkipListIndex:
+				if _, _, err := destinationColl.EnsureSkipListIndex(ctx, index.Fields(), &driver.EnsureSkipListIndexOptions{
+					Name:          index.UserName(),
+					Unique:        index.Unique(),
+					Sparse:        index.Sparse(),
+					NoDeduplicate: !index.Deduplicate(),
+				}); err != nil {
+					return err
+				}
+			case driver.HashIndex:
+				if _, _, err := destinationColl.EnsureHashIndex(ctx, index.Fields(), &driver.EnsureHashIndexOptions{
+					Name:          index.UserName(),
+					Unique:        index.Unique(),
+					Sparse:        index.Sparse(),
+					NoDeduplicate: !index.Deduplicate(),
+				}); err != nil {
+					return err
+				}
+			case driver.FullTextIndex:
+				if _, _, err := destinationColl.EnsureFullTextIndex(ctx, index.Fields(), &driver.EnsureFullTextIndexOptions{
+					Name:      index.UserName(),
+					MinLength: index.MinLength(),
+				}); err != nil {
+					return err
+				}
+			case driver.GeoIndex:
+				if _, _, err := destinationColl.EnsureGeoIndex(ctx, index.Fields(), &driver.EnsureGeoIndexOptions{
+					Name:    index.UserName(),
+					GeoJSON: index.GeoJSON(),
+				}); err != nil {
+					return err
+				}
+			case driver.EdgeIndex:
+				// These are automatically created.
+			case driver.PrimaryIndex:
+				// These are automatically created.
+			default:
+				return errors.New("unknown index type " + string(index.Type()))
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensureDestinationCollection ensures that a collection exists for a database on the destination and returns it.
+func (c *copier) ensureDestinationCollection(ctx context.Context, db driver.Database, coll driver.Collection, props driver.CollectionProperties) (driver.Collection, error) {
+	exists, err := db.CollectionExists(ctx, coll.Name())
+	if err != nil {
+		c.Logger.Warn().Err(err).Msg("Failed to get if collection exists.")
+		return nil, err
+	}
+	if !exists {
+		options := &driver.CreateCollectionOptions{
+			JournalSize:       int(props.JournalSize),
+			ReplicationFactor: props.ReplicationFactor,
+			WriteConcern:      props.WriteConcern,
+			WaitForSync:       props.WaitForSync,
+			DoCompact:         &props.DoCompact,
+			CacheEnabled:      &props.CacheEnabled,
+			ShardKeys:         props.ShardKeys,
+			NumberOfShards:    props.NumberOfShards,
+			IsSystem:          false,
+			Type:              props.Type,
+			KeyOptions: &driver.CollectionKeyOptions{
+				AllowUserKeys: props.KeyOptions.AllowUserKeys,
+				Type:          props.KeyOptions.Type,
+			},
+			DistributeShardsLike: props.DistributeShardsLike,
+			IsSmart:              false,
```

Comment:
Why not copy? Could be relevant for smart graphs.

Reply:
Stated in chat, but will copy here. IsSmart is missing from props. I'm working around that fact with a couple of lines below:
```go
		if props.SmartJoinAttribute != "" {
			options.IsSmart = true
			options.SmartJoinAttribute = props.SmartJoinAttribute
		}
```

I'll add the property in go-driver later on.